src/components/application_manager/src/commands/button_notification_to_mobile.cc

```diff
@@ -0,0 +1,231 @@
+    } else {
+      SDL_LOG_ERROR("Application " << app->app_id()
+                                   << " is not subscribed to button "
+                                   << btn_id);
+    }
+
+    return;
+  }
+
+  const auto subscribed_apps = SubscribedApps();
+  const auto app_ptr =
+      std::find_if(subscribed_apps.begin(),
+                   subscribed_apps.end(),
+                   [](const ApplicationSharedPtr subscribed_app) {
+                     return subscribed_app->IsFullscreen();
+                   });
+
+  if (app_ptr != subscribed_apps.end()) {
+    SDL_LOG_DEBUG("Sending button press for app in FULL with app id: "
+                  << (*app_ptr)->app_id());
+    SendButtonNotification(*app_ptr);
+    return;
+  } else {
+    SDL_LOG_ERROR("No application found");
+  }
+}
+
+std::vector<ApplicationSharedPtr> ButtonNotificationToMobile::SubscribedApps()
+    const {
+  const uint32_t btn_id = static_cast<uint32_t>(
+      (*message_)[strings::msg_params][hmi_response::button_name].asInt());
+  return application_manager_.applications_by_button(btn_id);
+}
+
+void ButtonNotificationToMobile::HandleMediaButton(
+    app_mngr::ApplicationSharedPtr app) {
+  SDL_LOG_AUTO_TRACE();
+  using namespace application_manager;
+
+  if (app) {
+    const auto btn_id = static_cast<mobile_apis::ButtonName::eType>(
+        (*message_)[strings::msg_params][hmi_response::button_name].asInt());
+
+    if (app->IsSubscribedToButton(btn_id)) {
+      SendButtonNotification(app);
+    } else {
+      SDL_LOG_ERROR("Application " << app->app_id()
+                                   << " is not subscribed to button "
+                                   << btn_id);
+    }
+
+    return;
+  }
+
+  const auto subscribed_apps = SubscribedApps();
+  const auto app_ptr =
+      std::find_if(subscribed_apps.begin(),
+                   subscribed_apps.end(),
+                   [](const ApplicationSharedPtr subscribed_app)
+
+                   {
+                     return helpers::Compare<mobile_api::HMILevel::eType,
+                                             helpers::EQ,
+                                             helpers::ONE>(
+                         subscribed_app->hmi_level(
+                             mobile_apis::PredefinedWindows::DEFAULT_WINDOW),
+                         mobile_api::HMILevel::HMI_FULL,
+                         mobile_api::HMILevel::HMI_LIMITED);
+                   }
+
+      );
+
+  if (app_ptr != subscribed_apps.end()) {
+    SendButtonNotification(*app_ptr);
+  } else {
+    SDL_LOG_ERROR("No application found");
+  }
+}
+
+void ButtonNotificationToMobile::Run() {
+  SDL_LOG_AUTO_TRACE();
+  using namespace application_manager::strings;
+
+  const uint32_t btn_id = static_cast<uint32_t>(
+      (*message_)[msg_params][hmi_response::button_name].asInt());
+
+  SDL_LOG_DEBUG("Received button id: " << btn_id);
+
+  ApplicationSharedPtr app;
+
+  if (DoesParamExist(strings::app_id)) {
```

Comment:
The check `(*message_).keyExists(msg_params)` can be removed from `DoesParamExist` because msg_params are already accessed on line 210 before any call of `DoesParamExist` method

Reply:
@iCollin agree, in all scenarios `msg_params` was accessed before. Removed in https://github.com/smartdevicelink/sdl_core/pull/3743/commits/daa1060833441f38a7219d8a852c6d44e852e76e